### ISSUE TITLE
allow for duplicate value labels in extractData with convertLabels = "factor"

### ISSUE THREAD

**Issue by weigtnik** [beckerbenj/eatGADS#77]
Suggestion: Append duplicate value labels during `char2fac` in `extractData` to prevent the function to abort with an error in the case of two (used) values of the same variable having the same label.

Specific use case: In ISCO-08, categories 82[00] and 821[0] (and probably others too) are both labeled "Assemblers" and both occur in the IQB-BT 2016 student data (SUF-Remote; variable `Eisco08f`). Executing `extractData(..., convertLabels = "factor")` results in an error due to the duplicate factor level.

**Comment by beckerbenj** [comment]
I had a short look into this:
Unfortunately, the extension as proposed by @NikklW is currently not possible within the `char2fac()` framework, as the function only receives the data as already transformed character vectors (argument `dat`).

Of course, the feature could be implemented via a larger overhaul of the function layout. However, I am not 100% sure I am in favor of the feature anyway. Alternatively, it could be argued that an informative error could help users modify the labels in the desired fashion before extracting the data.

What are your opinions on this @NikklW and @franikowsp? What was the use case that triggered this issue? Would an informative error imply substantial additional work?

**Comment by NikklW** [comment]
Thank you for pointing that out! I did not notice that `char2fac()` was the wrong spot to address the issue.

**Use case**: As mentioned briefly in the original post, the issue occurred with the student data from the [IQB BT 2016](http://doi.org/10.5159/IQB_BT_2016_v1) (version 1 was used in that case, however the issue is not specific to this version of the data). After `stata_datSuS <- import_stata()`, the data user tried to use `datSuS <- extractData(stata_datSuS, convertLabels = "factor")`, which ran into an error: "[...] factor level [427] is duplicated".
The issue lies with variable `Eisco08f`, which contains information on the paternal job coded in the International Standard Classification of Occupations, ISCO-08. It is caused by the fact that the classification sometimes uses the same label for a 2-digit category and the first of its 3-digit categories. In this example, 8200 and 8210 are each assigned to at least one case in the data and are both labeled as "Assemblers."

I think an informative error would help greatly because, right now, finding the cause of the issue is probably not trivial for most users. If I don't miss something important, this might be done with a relatively simple check during [`labels2values()`](https://github.com/beckerbenj/eatGADS/blob/86f1d7561152c2d30f6516da8c94c61051b2797e/R/extractData.R#L127), something like the following:

```
if (any(duplicated(change_table$value_new))) {
  duprows <- unique(c(rownames(change_table[duplicated(change_table$value_new, fromLast = FALSE),]),
                      rownames(change_table[duplicated(change_table$value_new, fromLast = TRUE),])))
  dupvars <- unique(change_table$varName[rownames(change_table) %in% duprows])
  stop("Duplicated value labels detected in variable(s) ", dupvars, ".",
       unlist(lapply(dupvars, FUN = function(var) paste0("\n affected values in variable ", var, ": ",
                                                  paste0(change_table$value[rownames(change_table) %in% duprows &
                                                                       change_table$varName == var],
                                                         collapse = ", ")))))
}
```
Just prettier, maybe.

**Comment by beckerbenj** [comment]
> the issue occurred with the student data from the [IQB BT 2016](http://doi.org/10.5159/IQB_BT_2016_v1)

I see! That is indeed unfortunate. There should not be any issues with the data sets we provide in combination with the software we recommend... 

> Thank you for pointing that out! I did not notice that char2fac() was the wrong spot to address the issue.

I would not call it the wrong spot, I think. It also depends on whether `extractData()` behavior should be modified for `convertLabels = "factor"` only or `convertLabels = "character"` as well. 

The current state is as follows:
- `convertLabels = "character"` silently recodes different values into the same value label 
- `convertLabels = "factor"` throws a hard to understand error message

Probably it would be desirable to have comparable behavior for `convertLabels = "character"` and `convertLabels = "factor"`? With the three options being:
- Recode different values into the same value label (with or without an informative warning)
- Auto-recode the value labels into differing value labels (with or without an informative warning)
- Throw an informative error

Intuitively, I personally would prefer option one (with an informative warning), as auto-recoding might require users to recode the resulting values again (into the same value). Furthermore, if users want different values, they probably will recode manually into self-defined value labels beforehand, wouldn't they?

**Comment by NikklW** [comment]
> Furthermore, if users want different values, they probably will recode manually into self-defined value labels beforehand, wouldn't they?
This is a very good point. So, since there are probably very few cases where this issue actually poses a problem, I agree that option 1 (with a warning) is likely the best choice.

**Comment by beckerbenj** [comment]
Sounds good! Will implement hopefully in the next week(s).

**Comment by beckerbenj** [comment]
Closed via #87
